### PR TITLE
Charles Ren - Mod 2 MidTerm

### PR DESCRIPTION
**Pull request by chuckrenny**
I think this mod has finally started clicking into place for me. I'm still not entirely comfortable writing code from scratch for some reason compared to Mod 1. I often have to reference syntax/naming conventions such as is it plural or not while I'm coding. Occasionally looking up ActiveRecord methods but knowing how to use them. I feel like I still need to solidify the second half of Week 2, Week 3 material came easier. Overall, just keep swimming.

### CODE REVIEW

File: app/controllers/customer_items_controller.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class CustomerItemsController < ApplicationController
+  def create
+    CustomerItems.create(customer_id: params[:customer_id], item_id: params[:item_id])
```
Comment (juliet-e) [inline app/controllers/customer_items_controller.rb:3]: Awesome set up here! This is the most restful option: putting this functionality in a joins controller `create` rather than a `create` in the CustomersController. Nice work!

File: app/models/customer.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class Customer < ApplicationRecord
+  validates_presence_of :name
+
+  has_and_belongs_to_many :items, join_table: 'customer_items'
```
Comment (juliet-e) [inline app/models/customer.rb:4]: What we've taught you so far with many-to-many-tables doesn't always work perfectly with this has_and_belongs_to_many association. That association is the same idea but if you use that association, your joins table can't have any additional fields besides the foreign key. We want the flexibility to be able to add columns to our joins tables as needed (for example, adding an additional status column to the PetApplications table in the last project). So, we teach students the has_many :items, through :customer_items instead of this association so that you have more flexibility.

tl;dr: while this association can be handy in some cases, use the association `has_many .. through:` instead because it's more flexible in all cases.

Comment (juliet-e) [reply app/models/customer.rb:4]: For this case though, this association works! But don't want you to try to use it in other situations where it won't work which is why we prefer `has_many .. through:`

File: app/models/item.rb [hunk 3]
```diff
@@ -1,3 +1,10 @@
 class Item < ApplicationRecord
+  validates_presence_of :name, :price
+
   belongs_to :supermarket
+  has_and_belongs_to_many :customers, join_table: 'customer_items'
+
+  def customer_count
+    self.customers.count
```
Comment (juliet-e) [inline app/models/item.rb:8]: Excellent work encapsulating this in a model method! Even though this is a simple data calculation, it's still a data calculation so to adhere to MVC strictly, this is the best place for it to be!

File: app/models/supermarket.rb [hunk 4]
```diff
@@ -1,3 +1,5 @@
 class Supermarket < ApplicationRecord
+  validates_presence_of :name, :location
+
   has_many :items
```
Comment (juliet-e) [inline app/models/supermarket.rb:4]: In order to be able to call `@supermarket.customers` in your controller, you need to add this association to directly link supermarkets to customers (they are already linked via supermarket -> items -> customer_items -> items but we want to tell AR to to do this DIRECTLy! Because this is a lengthy join otherwise.

How can we set up this direct link from supermarket to customers?: Set up the Supermarket model to take advantage of the associations that items, its child object has. Similar to how you use it with many-to-many relationships, you could add the following association to the Supermarket model:
`has_many :customers, through: :items`

Under the hood, AR will build an association that gets all the customers associated with the items that belong to this store. This is a multi-step join under the hood, but with AR, we can just call `supermarket.customers` and get the results!

File: app/views/customers/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+<ul class="items-list">
+  <% @customer.items.each do |item| %>
+    <p> Item Name: <%= item.name%> </p>
+    <p> Item Price: <%= item.price%> </p>
+    <p> Item Store: <%= item.supermarket.name%> </p>
+  <% end%>
+</ul>
+
+<%= form_with url: "/customers/#{@customer.id}/items" do |f| %>
```
Comment (juliet-e) [inline app/views/customers/show.html.erb:11]: Great, RESTful path choice here!

File: config/routes.rb [hunk 6]
```diff
@@ -3,4 +3,16 @@
 
   # Defines the root path route ("/")
   # root "articles#index"
+
```
Comment (juliet-e) [inline config/routes.rb:6]: Excellent RESTful routes!

File: spec/features/customers/show_spec.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      CustomerItems.create!(customer: @customer_2, item: @item_3)
+    end
+
+    # User Story 1
+    it "displays customer's name and list of customer's items with attributes" do
+      visit "/customers/#{@customer_1.id}"
+
+      expect(page).to have_content("Name: #{@customer_1.name}")
+
+      within(".items-list") do
```
Comment (juliet-e) [inline spec/features/customers/show_spec.rb:27]: Nice use of within blocks here! 

Another option: putting a html/css class on a div inside the iteration you do for the @customer.items.each
```
  <% @customer.items.each do |item| %>
    <div class = <% "item-#{item.id}" %>>
        <p>Item: <%= item.name %></p>
        <p>Cost: <%= item.price %></p>
        <p>Store: <%= item.supermarket.name %></p></br>
   </div>
 <% end %>
```

Now you have a div with a specific item ID so you can do `within(".item-#{item1.id}")` and get even more granular with your assertions!

Comment (juliet-e) [reply spec/features/customers/show_spec.rb:27]: We should also see expect statements here for item 2's attributes which are already associated to customer 1

File: spec/features/customers/show_spec.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    it "has a form to add an item to the customer" do
+      visit "/customers/#{@customer_1.id}"
+
+      expect(page).to_not have_content(@item_3.name)
+      fill_in :item_id, with: @item_3.id
+
+      click_button "Submit"
+
+      expect(current_path).to eq("/customers/#{@customer_1.id}")
+      expect(page).to have_content(@item_3.name)
```
Comment (juliet-e) [inline spec/features/customers/show_spec.rb:44]: Excellent, meaningful test for this form functionality

File: spec/features/items/index_html.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+      expect(page).to have_content("Name: #{@item_1.name}")
+      expect(page).to have_content("Price: #{@item_1.price}")
+      expect(page).to have_content("From: #{@item_1.supermarket.name}")
+    end
+
+    # User Story 3
+    it "displays the count of customers that bought the item" do
+      visit "/items"
+
+      expect(page).to have_content("Count: #{@item_1.customer_count}")
```
Comment (juliet-e) [inline spec/features/items/index_html.rb:34]: The structure of this test is excellent -- my only suggestion would be to not call the same `customer_count` method that you are testing. 

Consider the possibility that your logic in the `customer_count` method is somehow incorrect (it isn't, but let's imagine :) ). If you were to call this same method in the test, then your test would pass because it would be calling this incorrect method in order to verify that the method works. So, a more meaningful assertion is to do the calculation manually (it's pretty easy in this case - Item 1 is only associated with 1 customer) and use that result in the assertion directly: `expect(page).to have_content("Count: 1")`

That way, we can be sure that the functionality of the method is behaving correctly.

Comment (juliet-e) [reply spec/features/items/index_html.rb:34]: Since this is such a simple case bc there is only one customer, this is another reason why we always want to test these model methods in a model spec as well. It's a bit easier to set up data there so we can test more complex use cases (like when there is more than just one customer associated to an item)

File: spec/models/customer_spec.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+require 'rails_helper'
+
+RSpec.describe Customer, type: :model do
+  describe 'relationships' do
+    it { should have_and_belong_to_many :items }
+  end
+end
```
Comment (juliet-e) [inline spec/models/customer_spec.rb:7]: Just missing the model tests for your instance methods!

File: spec/models/item_spec.rb [hunk 11]
```diff
@@ -3,5 +3,6 @@
 RSpec.describe Item, type: :model do
   describe 'relationships' do
     it { should belong_to :supermarket }
+    it { should have_and_belong_to_many :customers }
   end
```
Comment (juliet-e) [inline spec/models/item_spec.rb:7]: Just missing the model tests for your instance methods!
